tools/skbstat.py

```diff
@@ -0,0 +1,1015 @@
+{
+	union filter_args args = { };
+
+	args.input = true;
+
+	return sk_match_and_trace(ctx, sk, &args, &record_sk);
+}
+
+static inline int __trace_skb_input(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.input = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+static inline int __trace_sk_output(struct pt_regs *ctx, struct sock *sk)
+{
+	union filter_args args = { };
+
+	args.output = true;
+
+	return sk_match_and_trace(ctx, sk, &args, &record_sk);
+}
+
+static inline int __trace_skb_output(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.output = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+int trace_sk_input(struct pt_regs *ctx, struct sock *sk)
+{
+	return __trace_sk_input(ctx, sk);
+}
+
+int trace_skb_input_arg3(struct pt_regs *ctx, void *arg1, void *arg2,
+			 struct sk_buff *skb)
+{
+	return __trace_skb_input(ctx, skb);
+}
+
+int trace_skb_input(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	return __trace_skb_input(ctx, skb);
+}
+
+int trace_sk_output(struct pt_regs *ctx, struct sock *sk)
+{
+	return __trace_sk_output(ctx, sk);
+}
+
+int trace_skb_output_arg3(struct pt_regs *ctx, void *arg1, void *arg2,
+			  struct sk_buff *skb)
+{
+	return __trace_skb_output(ctx, skb);
+}
+
+int trace_skb_output(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	return __trace_skb_output(ctx, skb);
+}
+
+int trace_skb_forward(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.forward = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+int trace_skb_drop(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.drop = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb_backtrace);
+}
+
+int trace_skb_consume(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb_backtrace);
+}
+"""
+
+# apply filters
+bpf_text = bpf_text.replace('SADDR_FILTER', saddr_filter)
+bpf_text = bpf_text.replace('DADDR_FILTER', daddr_filter)
+bpf_text = bpf_text.replace('SPORT_FILTER', sport_filter)
+bpf_text = bpf_text.replace('DPORT_FILTER', dport_filter)
+bpf_text = bpf_text.replace('PROTO_FILTER', proto_filter)
+bpf_text = bpf_text.replace('FLAGS_FILTER', flags_filter)
+bpf_text = bpf_text.replace('IP_VERSION_FILTER', ip_version_filter)
+
+# dump ebpf and exit
+if args.ebpf:
+    print(bpf_text)
+    exit()
+
+# initialize BPF
+b = BPF(text=bpf_text)
+
+# PROBE LIST
+# the dict key is function to probe, value is tuple.
+# the first entry of tuple is probe function, the second
+# one is the output priority, which can be used by sorter()
+# function. the name of probe function with biggest output 
+# priority number sits at last of output lines.
+
+# NOTE: those kernel functions may be inlined by compiler
+# or just not exists in older kernel. so it may lead to some 
```

Comment:
Let us not say "older kernel". It is totally possible that inlining or function renaming may happen in future kernels. Let us just mention "in some kernel versions". The same for below a couple of places.

Reply:
Yeah, make sense. not just older kernel.